mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/control/plane/resource/cluster/resourceprovider/SimpleFileResourceClusterStorageProvider.java

```diff
@@ -0,0 +1,170 @@
+
+                            return rcBuilder.cluster(
+                                            clusterSpecW.getId(),
+                                            RegisteredResourceClustersWritable.ClusterRegistration.builder()
+                                                    .clusterId(clusterSpecW.getId())
+                                                    .version(clusterSpecW.getVersion())
+                                                    .build())
+                                    .build();
+                        })
+                )
+                .mapAsync(1, rc -> updateRegisteredClusters(rc))
+                .mapAsync(1, rc -> updateClusterSpecImpl(clusterSpecWritable))
+                .mapAsync(1, rc -> getResourceClusterSpecWritable(rc.getId()))
+                .runWith(Sink.last(), system);
+        logger.info("Return future on registerAndUpdateClusterSpec: " + clusterSpecWritable.getId());
+        return fut;
+    }
+
+    public CompletionStage<ResourceClusterSpecWritable> updateClusterSpecImpl(ResourceClusterSpecWritable spec) {
+
+        Sink<ByteString, CompletionStage<IOResult>> fileSink = FileIO.toPath(
+                getClusterSpecFilePath(SPOOL_DIR, spec.getId()));
+        Source<ResourceClusterSpecWritable, NotUsed> textSource = Source.single(spec);
+
+        return textSource
+                .map(w -> mapper.writeValueAsString(w))
+                .map(ByteString::fromString)
+                .runWith(fileSink, system)
+                .exceptionally(e -> { throw new RuntimeException("failed to save spec: " + spec.getId(), e); })
+                .thenComposeAsync(ioRes -> getResourceClusterSpecWritable(spec.getId())); // IOResult result is always true.
+    }
+
+    public CompletionStage<Boolean> updateRegisteredClusters(RegisteredResourceClustersWritable clusters) throws IOException {
+        Path listFilePath = getClusterListFilePath();
+        if (Files.notExists(listFilePath)) {
+            Files.createDirectories(Paths.get(SPOOL_DIR));
+            Files.createFile(listFilePath);
+        }
+
+        Sink<ByteString, CompletionStage<IOResult>> listFileSink = FileIO.toPath(
+                getClusterListFilePath());
+        Source<RegisteredResourceClustersWritable, NotUsed> textSource = Source.single(clusters);
+        return
+                textSource
+                        .map(mapper::writeValueAsString)
+                        .map(ByteString::fromString)
+                        .runWith(listFileSink, system)
+                        .exceptionally(e -> { throw new RuntimeException("failed to save cluster list. ", e); })
+                        .thenApply(ioRes -> true); // IOResult result is always true.
+    }
+
+    @Override
+    public CompletionStage<RegisteredResourceClustersWritable> getRegisteredResourceClustersWritable() {
+        final ObjectMapper mapper = new ObjectMapper();
```

Comment:
I don't believe you need a separate instance of `ObjectMapper` here since `readerFor` and `ObjectReader`(s) look thread safe anyway.